### DISCUSSION TITLE
%PhantomCamera3D can't be found

### DISCUSSION THREAD

**Discussion by suau** [ramokz/phantom-camera#188] [Question / Help Wanted]
Hi,
I've added PhantomCamera3D using ThirdPerson Follow Mode.
Even though I've marked it as % to be able to get a referece to it anywhere, it can't be found this way.
Upon debugging, it seems to get re-created as a child-node of a `SpringArm3D` an thus loosing the unique handle.

Just for context, when respawning the player, I'd like to call `set_follow_target_node` on it.
As a workaround I'm currently using `$Camera3D/PhantomCameraHost.get_active_pcam()` to get a hold of it and re-assign it, which works for me, but it could be an issue if multiple cameras are involved, like in the 3DFollowThirdPersonExampleScene

**Comment by ramokz** [comment]
That's strange, can't seem to replicate it in the example scene at least.

Are you referencing it directly via `%PhantomCamera`, or are you creating a variable using `@onready var`?
Also, thinking it might be an instantiation order / load situation, how many nodes does your scene roughly have?

**Reply by suau** [reply]
@ramokz very few nodes, I'm new to Godot and just testing things out on a small platform.

I did some more testing and believe this is probably a Godot bug, which seems to break unique-name search after reparenting.

1) First I confirmed the unique-name search doesn't work anymore:

I added this line to the `_physics_process:` function of the `3DFollowThirdPersonExampleScene`:
`print("%PlayerPhantomCamera3D: ", %PlayerPhantomCamera3D)` and it turns to `null` after a few calls.

2) Then I confirmed the object-id pcam_host sees remains the same:
<img width="1477" alt="Screenshot 2024-01-23 at 12 38 29" src="https://github.com/ramokz/phantom-camera/assets/2596700/113dac8b-530e-415e-98ba-6876b68dc1c8">

3) Then I added some lines here to see if it breaks at the time I suspect:
https://github.com/ramokz/phantom-camera/blob/5419a782de66095123792a67e07da1a54b63a0ed/addons/phantom_camera/scripts/phantom_camera/phantom_camera_3D.gd#L666
before and after reparenting.
```
print("REPARENTING BEFORE name: ", name)
print("REPARENTING BEFORE unique: ", has_unique_name)
reparent(_follow_spring_arm_node)
print("REPARENTING AFTER name: ", name)
print("REPARENTING AFTER unique: ", has_unique_name)
```

And exactly after reparenting it can't be found anymore, although it will continues to have the `has_unique_name` be `true` , I double-confirmed this in `_physics_process` using the `@onready var` to make sure it's not a timing/deferred issue.
Just to be clear, the `@onready` remains accessible, but the getting the node using a get_node search with `%PlayerPhantomCamera3D` doesn't.

**Comment by ZenithStar** [comment]
I've also encountered this issue before, specifically for third person follow. Reparenting to the spring arm breaks the unique name reference. As a workaround, I have my main world manager grab and hold the PhantomCamera3D node reference before giving it a follow target.

**Reply by suau** [reply]
That'll work !
